From sarah.a.sharp@linux.intel.com  Fri May  1 15:02:44 2009
From: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Date: Wed, 29 Apr 2009 19:05:40 -0700
Subject: USB: xhci: Clean up xhci_irq() function.
To: Greg KH <greg@kroah.com>
Message-ID: <20090430020540.GA31212@gamba.jf.intel.com>


Drop spinlock in xhci_irq() error path.
This fixes the issue reported by Oliver Neukum on this thread:
	http://marc.info/?l=linux-usb&m=124090924401444&w=2

Remove unnecessary register read reported by Viral Mehta:
	http://marc.info/?l=linux-usb&m=124091326007398&w=2

Reported-by: Oliver Neukum <oliver@neukum.org>
Reported-by: Viral Mehta <viral.mehta@einfochips.com>
Signed-off-by: Sarah Sharp <sarah.a.sharp@linux.intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/host/xhci-hcd.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/usb/host/xhci-hcd.c
+++ b/drivers/usb/host/xhci-hcd.c
@@ -276,11 +276,11 @@ irqreturn_t xhci_irq(struct usb_hcd *hcd
 		return IRQ_NONE;
 	}
 
-	temp = xhci_readl(xhci, &xhci->op_regs->status);
 	if (temp & STS_FATAL) {
 		xhci_warn(xhci, "WARNING: Host System Error\n");
 		xhci_halt(xhci);
 		xhci_to_hcd(xhci)->state = HC_STATE_HALT;
+		spin_unlock(&xhci->lock);
 		return -ESHUTDOWN;
 	}
 
